Add acceleration input to TrajectoryCar and enable AutoDiffXd

jadecastro opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6119)
<!-- Reviewable:end -->

jadecastro commented:
+@liangfok for feature review, please

---

Review status: 0 of 21 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-:-KkI-4aySpn4kgO3H5UR:bpc22g)*
<!-- Sent from Reviewable.io -->

liangfok commented:
First pass completed.

---

Reviewed 22 of 22 files at r1.
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2ekqHdMlEBOk3Bak:-KkM2ekqHdMlEBOk3Bal:b-o5aluf) (no related file):*
BTW, instead of modifying `TrajectoryCar` to be a `PathFollowingCar`, would it be better to have both in-tree? Having a time-predictable model like `TrajectoryCar` may be useful.

---

*[drake/automotive/curve2.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM3D5cxNS14jryzfFs:-KkM3D5cxNS14jryzfFt:bouxsom) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/curve2.cc#L3)):*
> ```cpp
> #include "drake/automotive/curve2.h"
> 
> // #include "drake/common/autodiff_overloads.h"
> ```

Remove commented out code.

---

*[drake/automotive/trajectory_car.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM3MMUErkgXWN-DFtG:-KkM3MMUErkgXWN-DFtH:b57zmzs) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.cc#L3)):*
> ```cpp
> #include "drake/automotive/trajectory_car.h"
> 
> // #include "drake/common/autodiff_overloads.h"
> ```

Removed commented out code.

---

*[drake/automotive/trajectory_car.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwAW94AS6tsxWHf94:-KkLwAW94AS6tsxWHf95:b-6rxzob) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L33)):*
> ```objc
> /// * desired acceleration, a systems::BasicVector of size 1 (optional input).
> ///   If left unconnected, the trajectory car will travel at the initial
> ///   velocity, specifed via the `speed` input argument.
> ```

`speed` should be `initial_speed`. This sentence should also indicate that it's referring to the constructor's argument list.

---

*[drake/automotive/trajectory_car.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwNrvcZZyw95QgTlN:-KkLwNrvcZZyw95QgTlO:bdq565h) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L33)):*
> ```objc
> /// * desired acceleration, a systems::BasicVector of size 1 (optional input).
> ///   If left unconnected, the trajectory car will travel at the initial
> ///   velocity, specifed via the `speed` input argument.
> ```

`velocity` should be `speed` because the heading will change as the vehicle follows the curve.

---

*[drake/automotive/trajectory_car.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLvi4c0ZQNe-wX7O-w:-KkLvi4c0ZQNe-wX7O-x:b-rwz10w) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L48)):*
> ```objc
> ///   (OutputPortDescriptor getter: velocity_output())
> ///
> /// @ingroup automotive_plants
> ```

BTW, consider adding `TrajectoryCarParams` to this documentation.

---

*[drake/automotive/trajectory_car.h, line 57 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLvrkSvIPXQ27hrxJl:-KkLvrkSvIPXQ27hrxJm:bxff0kd) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L57)):*
> ```objc
> 
>   /// Constructs a TrajectoryCar system that traces the given @p curve, at the
>   /// given constant @p speed, starting at the given @p start_position - the
> ```

"at the given constant @p speed" needs to be updated.

---

*[drake/automotive/trajectory_car.h, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwavNLcC64jPpPorg:-KkLwavOTcJL6J4Iqzk3:b374hbw) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L58)):*
> ```objc
>   /// Constructs a TrajectoryCar system that traces the given @p curve, at the
>   /// given constant @p speed, starting at the given @p start_position - the
>   /// initial position along the (possibly extrapolated) curve.  Throws an error
> ```

Can you provide more details about how the curve is extrapolated to allow arbitrary `start_position` parameters? Are arbitrary `start_position` values supported?

---

*[drake/automotive/trajectory_car.h, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwwXb8fPxjGhZufql:-KkLwwXb8fPxjGhZufqm:bj4tom1) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L59)):*
> ```objc
>   /// given constant @p speed, starting at the given @p start_position - the
>   /// initial position along the (possibly extrapolated) curve.  Throws an error
>   /// if the curve is empty (a zero @p path_length).
> ```

Will this also throw an error if `initial_speed` is beyond the maximum speed specified by `TrajectoryCarParams`.

---

*[drake/automotive/trajectory_car.h, line 63 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM25yAXt2KuUaTrGmF:-KkM25yAXt2KuUaTrGmG:b-ulnzyh) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L63)):*
> ```objc
>       : curve_(curve),
>         initial_speed_(initial_speed),
>         start_position_(start_position) {
> ```

Instead of saving `initial_speed` and `start_position` in member variables, why not directly store them in `TrajectoryCarState`?

---

*[drake/automotive/trajectory_car.h, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM0ZT7_gaEaytqsDXc:-KkM0ZT7_gaEaytqsDXd:b-offba1) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L145)):*
> ```objc
>         this->template EvalVectorInput<systems::BasicVector>(context, 0);
> 
>     // If the input is null, then apply a default acceleration setting of zero.
> ```

BTW, consider replacing "acceleration setting" with "desired acceleration" for consistency.

---

*[drake/automotive/trajectory_car.h, line 224 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2JlscFMSoyyImCHc:-KkM2JlscFMSoyyImCHd:bzviq8) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L224)):*
> ```objc
>   }
> 
>   void SetDefaultState(const systems::Context<T>&,
> ```

This method is not used.

---

*[drake/automotive/trajectory_car.h, line 233 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM-ibyRgYkMaS7pbnP:-KkM-ibyRgYkMaS7pbnQ:b-l0tg7d) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L233)):*
> ```objc
>   }
> 
>   const PositionHeading CalcRawPose(const TrajectoryCarState<T>& state) const {
> ```

Please add a docstring for this method. In particular, I'm not sure what "raw" means in this context.

---

*[drake/automotive/trajectory_car.h, line 238 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2RAi7x_uu4Kuu-X9:-KkM2RAi7x_uu4Kuu-XA:b-rix32z) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L238)):*
> ```objc
>     PositionHeading result;
> 
>     // Compute the curve at the current longitudinal position.
> ```

For consistency, consider replacing "logitudinal" with "on-curve".

---

*[drake/automotive/trajectory_car_params.named_vector, line 2 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLx8pIaI6N3wO4V1Ds:-KkLx8pJnpqVOjXYlOgG:b-gel03q) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car_params.named_vector#L2)):*
> ```
> element {
>     name: "max_velocity"
> ```

Velocity typically includes both heading and speed. Since this is in units of m/s, consider calling this "speed".

(Same comment below regarding `velocity_limit_kp`.)

---

*[drake/automotive/trajectory_car_state.named_vector, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLxPHkqYfSmeLT23Kr:-KkLxPHkqYfSmeLT23Ks:bzau0eo) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car_state.named_vector#L3)):*
> ```
> element {
>     name: "position"
>     doc: "The along-path position of the vehicle."
> ```

Should "path" be "curve" for consistency?

(Same comment below.)

---

*[drake/automotive/test/trajectory_car_test.cc, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM4JMlEEO4Rp2odluR:-KkM4JMlEEO4Rp2odluS:b-shhk1z) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L39)):*
> ```cpp
> 
> // Check the car's progress along some simple paths.  We just want to test our
> // ImplCalcFoo() methods.  We can assume that Curve2 is correct, because has its
> ```

Missing "it" before "has".

---

*[drake/automotive/test/trajectory_car_test.cc, line 85 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM4UMcopbgx6K55nlW:-KkM4UMcopbgx6K55nlX:bg5v6eg) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L85)):*
> ```cpp
>     const TrajectoryCar<double> car_dut{curve, it.speed, it.start_position};
> 
>     // Check that the systems' outputs over time are correct over the
> ```

BTW, consider deleting "over time".

---

*[drake/automotive/test/trajectory_car_test.cc, line 159 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM6AQTToc595VXhWms:-KkM6AQUmDdSQ_v6GUxm:bz9x8y0) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L159)):*
> ```cpp
> }
> 
> // Tests the derivatives when a non-zero acceleration is provided as an input.
> ```

This test is also evaluating speed saturation when the desired acceleration is zero. Consider splitting into two unit tests.

---

*[drake/automotive/test/trajectory_car_test.cc, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5-hR_zxBQjroA576:-KkM5-hR_zxBQjroA577:btkm9kt) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L180)):*
> ```cpp
> 
>   // Create some non-saturating acceleration command.
>   double kAccelerationInput = 1.5;
> ```

This variable is not a constant and thus can't use the `kFoo` naming scheme.

---

*[drake/automotive/test/trajectory_car_test.cc, line 191 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM56h8FBCaYGft8cxI:-KkM56h8FBCaYGft8cxJ:b-mb454a) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L191)):*
> ```cpp
> 
>   // Set the speed to some saturated value.
>   double kNewSpeed = default_params.max_velocity() + 5.;
> ```

`const`.

---

*[drake/automotive/test/trajectory_car_test.cc, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5m4vjl3xGAWnPnKb:-KkM5m4vjl3xGAWnPnKc:b-lyj0kj) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L192)):*
> ```cpp
>   // Set the speed to some saturated value.
>   double kNewSpeed = default_params.max_velocity() + 5.;
>   car_state->set_speed(kNewSpeed);
> ```

BTW, is this even legal (i.e., should an exception be thrown)?

(Same comment regarding setting a negative speed.)

---

*[drake/automotive/test/trajectory_car_test.cc, line 197 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5D7ICt-4x0RIQ1om:-KkM5D7ICt-4x0RIQ1on:bfnfva8) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L197)):*
> ```cpp
>   car_dut.CalcTimeDerivatives(*context, derivatives.get());
> 
>   // Tests that the derivatives do in fact reflect that the acceleration has
> ```

I believe "acceleration" should be "speed". Since the speed is saturated, the car's acceleration should be negative.

---

*[drake/automotive/test/trajectory_car_test.cc, line 220 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM6RqAaJNsh2BnlMLh:-KkM6RqAaJNsh2BnlMLi:b-mtfacm) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L220)):*
> ```cpp
>   const double kInitialSpeed = 20.;
>   const TrajectoryCar<double> car_dut{curve, kInitialSpeed,
>                                       0. /* start s-position */};
> ```

"s-position" doesn't seem correct since "s" is a Maliput concept. Should it be "path position" instead?

---

*[drake/automotive/test/trajectory_car_test.cc, line 234 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM6atWKG-jwLkItyMU:-KkM6atWKG-jwLkItyMV:bj92r70) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L234)):*
> ```cpp
>   // Check that the public methods can be called without exceptions.
>   autodiff_dut->CalcOutput(*context, output.get());
>   autodiff_dut->CalcTimeDerivatives(*context, derivatives.get());
> ```

Should these use `EXPECT_NO_THROW()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-:-KkM3i-regaGi0GAZsov:b-nxgybn)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 10 of 25 files reviewed at latest revision, 25 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2ekqHdMlEBOk3Bak:-KkMKqd8P-G-Bq-e9P14:bxmaxh6) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, instead of modifying `TrajectoryCar` to be a `PathFollowingCar`, would it be better to have both in-tree? Having a time-predictable model like `TrajectoryCar` may be useful.
</blockquote></details>

That's a reasonable argument. Do you see any limitation of the proposed implementation to the existing use-case of TrajectoryCar? There is not real substantive change in this version other than the fact that it now makes sense to reason about the curve as a position-indexed curve rather than a time-indexed curve. Since the speed was constant in the original implementation, the behavior of the two versions are essentially the same (since the designer can achieve this by choosing to zero out the acceleration in the new implementation or simply not wire it up). 

One other thing to note (which might not be apparent here) is that, as a side effect of specifying a starting position (rather than a starting time), is that in the automotive simulator the cars all begin spaced evenly with respect to one another, rather than being on top of one another to start with, but with staggered start times. I feel that having cars not starting in collision with one another should be the norm, but am open to suggestions. 

---

*[drake/automotive/curve2.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM3D5cxNS14jryzfFs:-KkMNXyC5RQolIESTMmD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/curve2.cc#L3)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Remove commented out code.
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM3MMUErkgXWN-DFtG:-KkMNYXBsxKEtXWM2Sa1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.cc#L3)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Removed commented out code.
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwAW94AS6tsxWHf94:-KkMPVFWwVfnSoMLVHd3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L33)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`speed` should be `initial_speed`. This sentence should also indicate that it's referring to the constructor's argument list.
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwNrvcZZyw95QgTlN:-KkMPoBpXcD4ryFhesmi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L33)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`velocity` should be `speed` because the heading will change as the vehicle follows the curve.
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLvi4c0ZQNe-wX7O-w:-KkN9XtqrT7n3Ag3MZSg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L48)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding `TrajectoryCarParams` to this documentation.
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 57 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLvrkSvIPXQ27hrxJl:-KkMNaY_ZOh_QClYpUmL:b2ar3so) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L57)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"at the given constant @p speed" needs to be updated.
</blockquote></details>

Done. (removed)

---

*[drake/automotive/trajectory_car.h, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwavNLcC64jPpPorg:-KkMTJ2PW22Zl3RTuhpE:b8npgmo) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L58)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Can you provide more details about how the curve is extrapolated to allow arbitrary `start_position` parameters? Are arbitrary `start_position` values supported?
</blockquote></details>

I've removed the comment as per comment below to remove the two ctor arguments, replacing them with initial states.

---

*[drake/automotive/trajectory_car.h, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLwwXb8fPxjGhZufql:-KkMj6q8bQwDLAPz6MF4:bsqnp74) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L59)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Will this also throw an error if `initial_speed` is beyond the maximum speed specified by `TrajectoryCarParams`.
</blockquote></details>

It's simply clipped at `max_velocity`.

---

*[drake/automotive/trajectory_car.h, line 63 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM25yAXt2KuUaTrGmF:-KkMawRgJrXVxGVeaU1x:b-pz2px5) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L63)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Instead of saving `initial_speed` and `start_position` in member variables, why not directly store them in `TrajectoryCarState`?
</blockquote></details>

You are right, this is a better way to do that. I have updated the PR to include that and have removed them from the ctor argument list.

---

*[drake/automotive/trajectory_car.h, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM0ZT7_gaEaytqsDXc:-KkMyB6cVfOTwhUD8JT-:b6kog4b) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L145)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider replacing "acceleration setting" with "desired acceleration" for consistency.
</blockquote></details>

I replaced `acceleration setting` with `acceleration`, but I think `desired acceleration` gets into the weeds of `ImplCalcTimeDerivatives` and anyway are unaffected by the implementation (regardless of choice of `TrajectoryCarParams`). I've added a comment within `ImplCalcTimeDerivatives` that clarifies that zero is passed through unaffected.

---

*[drake/automotive/trajectory_car.h, line 224 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2JlscFMSoyyImCHc:-KkMz3j_Fs4CfAzBXahK:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L224)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This method is not used.
</blockquote></details>

Removed.

---

*[drake/automotive/trajectory_car.h, line 233 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM-ibyRgYkMaS7pbnP:-KkMz4uJddA2fb_6-XGB:b8qtml1) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L233)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Please add a docstring for this method. In particular, I'm not sure what "raw" means in this context.
</blockquote></details>

Done.

The "raw" in `raw_pose` is meant to imply that something needs to be done to the pose data that is returned by the function in order to render it usable.  In this specific context, it means that that `PoseHeading` has not yet been converted into the `PoseVector` -- that is done inside `ImplCalcPose`. 

---

*[drake/automotive/trajectory_car.h, line 238 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2RAi7x_uu4Kuu-X9:-KkN0QKO1FD3NlHVWLNs:b-ahs680) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L238)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

For consistency, consider replacing "logitudinal" with "on-curve".
</blockquote></details>

Done.  I've clarified the "longitudinal" qualifier.

---

*[drake/automotive/trajectory_car_params.named_vector, line 2 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLx8pIaI6N3wO4V1Ds:-KkN0bfUxPqgZuEfCyQH:bz2tm66) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car_params.named_vector#L2)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Velocity typically includes both heading and speed. Since this is in units of m/s, consider calling this "speed".

(Same comment below regarding `velocity_limit_kp`.)
</blockquote></details>

Done.  Note that this misuse also exists in `SimpleCarState` and `SimpleCarParams`.

---

*[drake/automotive/trajectory_car_state.named_vector, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkLxPHkqYfSmeLT23Kr:-KkN1R4pOx0cVzevXFoe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car_state.named_vector#L3)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should "path" be "curve" for consistency?

(Same comment below.)
</blockquote></details>

Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM4JMlEEO4Rp2odluR:-KkN2R1xsBlL3EUPu0Gm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L39)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Missing "it" before "has".
</blockquote></details>

Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 85 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM4UMcopbgx6K55nlW:-KkN2VppxW0GrEXtprSL:bihj9zs) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L85)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider deleting "over time".
</blockquote></details>

But, we are still simulating the system over time, so think it's still accurate?

---

*[drake/automotive/test/trajectory_car_test.cc, line 159 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM6AQTToc595VXhWms:-KkN4GlCklUHaodEONrF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L159)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This test is also evaluating speed saturation when the desired acceleration is zero. Consider splitting into two unit tests.
</blockquote></details>

Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5-hR_zxBQjroA576:-KkN383ZK4gRk7KmLqpE:bs7wumx) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L180)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This variable is not a constant and thus can't use the `kFoo` naming scheme.
</blockquote></details>

Done. 

---

*[drake/automotive/test/trajectory_car_test.cc, line 191 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM56h8FBCaYGft8cxI:-KkN7_xWd84ppOdhIHMK:b3yfxoq) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L191)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

`const`.
</blockquote></details>

(It is redefined below). 
Both sub-tests now have unique names for the speeds.

---

*[drake/automotive/test/trajectory_car_test.cc, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5m4vjl3xGAWnPnKb:-KkN8DosftSNKnLSxVZo:b-5ueuds) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L192)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, is this even legal (i.e., should an exception be thrown)?

(Same comment regarding setting a negative speed.)
</blockquote></details>

Why should an exception be thrown? The idea is to saturate the speed.

---

*[drake/automotive/test/trajectory_car_test.cc, line 197 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5D7ICt-4x0RIQ1om:-KkN8ROVUGEwscK4hlZh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L197)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I believe "acceleration" should be "speed". Since the speed is saturated, the car's acceleration should be negative.
</blockquote></details>

Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 220 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM6RqAaJNsh2BnlMLh:-KkN8lX6dc9I_mTCw0Fc:b-o724ak) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L220)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"s-position" doesn't seem correct since "s" is a Maliput concept. Should it be "path position" instead?
</blockquote></details>

Done.  (removed).

---

*[drake/automotive/test/trajectory_car_test.cc, line 234 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM6atWKG-jwLkItyMU:-KkN93JRgMz5z7kcSH_q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L234)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should these use `EXPECT_NO_THROW()`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 15 of 15 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM2ekqHdMlEBOk3Bak:-KkQmkQWnOHUJUSdQcBm:b-kw04sg) (no related file):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

That's a reasonable argument. Do you see any limitation of the proposed implementation to the existing use-case of TrajectoryCar? There is not real substantive change in this version other than the fact that it now makes sense to reason about the curve as a position-indexed curve rather than a time-indexed curve. Since the speed was constant in the original implementation, the behavior of the two versions are essentially the same (since the designer can achieve this by choosing to zero out the acceleration in the new implementation or simply not wire it up). 

One other thing to note (which might not be apparent here) is that, as a side effect of specifying a starting position (rather than a starting time), is that in the automotive simulator the cars all begin spaced evenly with respect to one another, rather than being on top of one another to start with, but with staggered start times. I feel that having cars not starting in collision with one another should be the norm, but am open to suggestions. 
</blockquote></details>

I accept your arguments and will mark myself satisfied. Yes, I agree the cars should be started in non-collision configurations and appreciate how the changes in this PR enable that.

---

*[drake/automotive/automotive_simulator.cc, line 223 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkQnyjDLPb2wztaua9r:-KkQnyjDLPb2wztaua9s:b16xkk) ([raw file](https://github.com/robotlocomotion/drake/blob/b899ea4e75a4fd7cc940bbb3ca13fa10e459ab4c/drake/automotive/automotive_simulator.cc#L223)):*
> ```cpp
>   vehicles_[id] = trajectory_car;
> 
>   TrajectoryCarState<double> initial_state{};
> ```

BTW, the mustache is not necessary since this is not a POD type.

---

*[drake/automotive/trajectory_car.h, line 233 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM-ibyRgYkMaS7pbnP:-KkQpGh6rJ5bnwxNyZkL:b5eb9lp) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L233)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Done.

The "raw" in `raw_pose` is meant to imply that something needs to be done to the pose data that is returned by the function in order to render it usable.  In this specific context, it means that that `PoseHeading` has not yet been converted into the `PoseVector` -- that is done inside `ImplCalcPose`. 
</blockquote></details>

BTW, `CalcRawPose()` is called by all three of the following methods:

* `ImplCalcOutput()`
* `ImplCalcPose()`
* `ImplCalcVelocity()`

Consider calling it once within `DoCalcOutput()` and passing a `const PositionHeading&` into the three above-mentioned methods.

---

*[drake/automotive/test/trajectory_car_test.cc, line 85 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM4UMcopbgx6K55nlW:-KkQq6JHQH5Cu6Gy9zmm:b-cj1uv3) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L85)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

But, we are still simulating the system over time, so think it's still accurate?
</blockquote></details>

Yes, it's accurate. I was just thinking that "over time" is redundant with "over the entire duration of the trajectory". I'll leave it at your discretion.

---

*[drake/automotive/test/trajectory_car_test.cc, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5m4vjl3xGAWnPnKb:-KkQqYJRLtgZOUgD3lvP:bkr6mci) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L192)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Why should an exception be thrown? The idea is to saturate the speed.
</blockquote></details>

Isn't `car_state` the actual speed of the vehicle, not the desired speed, meaning saturation occurs before the speed is saved into `car_state`?

---

*[drake/automotive/test/trajectory_car_test.cc, line 185 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkRERgePxiacgdaBlms:-KkRERgfLqi9C5xZNHri:b-ceegqo) ([raw file](https://github.com/robotlocomotion/drake/blob/b899ea4e75a4fd7cc940bbb3ca13fa10e459ab4c/drake/automotive/test/trajectory_car_test.cc#L185)):*
> ```cpp
>       context->get_mutable_continuous_state_vector());
>   ASSERT_NE(nullptr, car_state);
>   car_state->set_position(0.);
> ```

BTW, to be consistent with `kInitialSpeed`, should we also define a `kInitialPosition` and use it here?

---

*[drake/automotive/test/trajectory_car_test.cc, line 206 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkREfVdUZs7S_Ow8-iO:-KkREfVdUZs7S_Ow8-iP:b-j6e2v5) ([raw file](https://github.com/robotlocomotion/drake/blob/b899ea4e75a4fd7cc940bbb3ca13fa10e459ab4c/drake/automotive/test/trajectory_car_test.cc#L206)):*
> ```cpp
>                                        {10., 30.}};
>   const Curve2d curve{waypoints};
>   const TrajectoryCar<double> car_dut{curve};  // The device under test.
> ```

BTW, why use "car_dut" instead of the simpler "dut"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Second pass completed.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-:-KkRF8kHvhLEa_VJ7jAN:b9zdyj)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 22 of 25 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 223 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkQnyjDLPb2wztaua9r:-KkRds0h_-neLwIpma6d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b899ea4e75a4fd7cc940bbb3ca13fa10e459ab4c/drake/automotive/automotive_simulator.cc#L223)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, the mustache is not necessary since this is not a POD type.
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 233 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM-ibyRgYkMaS7pbnP:-KkRjHeqVkiN8_c250wZ:bis3nva) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L233)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, `CalcRawPose()` is called by all three of the following methods:

* `ImplCalcOutput()`
* `ImplCalcPose()`
* `ImplCalcVelocity()`

Consider calling it once within `DoCalcOutput()` and passing a `const PositionHeading&` into the three above-mentioned methods.
</blockquote></details>

Sorry for the confusion. The current implementation already has `CalcRawPose()` called once within `DoCalcOutput()`. (see: https://github.com/RobotLocomotion/drake/pull/6119/files#diff-f83cc49357c38626ef557fc12e109d7bR106)

---

*[drake/automotive/test/trajectory_car_test.cc, line 85 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM4UMcopbgx6K55nlW:-KkRe4qP8bND1We9rUpP:ba4l7la) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L85)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, it's accurate. I was just thinking that "over time" is redundant with "over the entire duration of the trajectory". I'll leave it at your discretion.
</blockquote></details>

I see what you mean now, thanks. Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5m4vjl3xGAWnPnKb:-KkTB_64AbA_1pdGEbBR:b5cyhkg) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L192)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Isn't `car_state` the actual speed of the vehicle, not the desired speed, meaning saturation occurs before the speed is saved into `car_state`?
</blockquote></details>

Let me rephrase - on speed there is an upper limit (`params.max_speed()`) and lower limit (zero). The the extent to which we are violating the upper limit drives how much the input command acceleration is overridden by a negative acceleration. This is done using the `calc_smooth_acceleration()` function in essentially the same way as in `SimpleCar`.  So, no exception is thrown.  I've clarified with a comment - hopefully that clarifies?

---

*[drake/automotive/test/trajectory_car_test.cc, line 185 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkRERgePxiacgdaBlms:-KkTAm9m_A6YRCd42Cbb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b899ea4e75a4fd7cc940bbb3ca13fa10e459ab4c/drake/automotive/test/trajectory_car_test.cc#L185)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, to be consistent with `kInitialSpeed`, should we also define a `kInitialPosition` and use it here?
</blockquote></details>

Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 206 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkREfVdUZs7S_Ow8-iO:-KkTB8MncqUFOLNSPz05:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b899ea4e75a4fd7cc940bbb3ca13fa10e459ab4c/drake/automotive/test/trajectory_car_test.cc#L206)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, why use "car_dut" instead of the simpler "dut"?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/trajectory_car.h, line 233 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM-ibyRgYkMaS7pbnP:-KkTVAd66yTuNXNIKpFj:b8tgi9y) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/trajectory_car.h#L233)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Sorry for the confusion. The current implementation already has `CalcRawPose()` called once within `DoCalcOutput()`. (see: https://github.com/RobotLocomotion/drake/pull/6119/files#diff-f83cc49357c38626ef557fc12e109d7bR106)
</blockquote></details>

Oops sorry. I don't know what I was looking at previously. Marking myself satisfied.

---

*[drake/automotive/test/trajectory_car_test.cc, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkM5m4vjl3xGAWnPnKb:-KkTVSXi1Z3eUTLIjMD_:btydw2e) ([raw file](https://github.com/robotlocomotion/drake/blob/ab3a5902d03bbc6d28de2d6834781f55a57c5595/drake/automotive/test/trajectory_car_test.cc#L192)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Let me rephrase - on speed there is an upper limit (`params.max_speed()`) and lower limit (zero). The the extent to which we are violating the upper limit drives how much the input command acceleration is overridden by a negative acceleration. This is done using the `calc_smooth_acceleration()` function in essentially the same way as in `SimpleCar`.  So, no exception is thrown.  I've clarified with a comment - hopefully that clarifies?
</blockquote></details>

Ah, I now see that the speed limit can, in fact, be violated. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-:-KkTVgthjMEau-HMqLqE:b593w57)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
+@soonho-tri for platform review, please

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-:-KkUqW8Dr5wdzO1fSXha:b-5s300q)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (aside suggestions for better documentation and optimization).

---

Reviewed 10 of 22 files at r1, 12 of 15 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWuJ5ifRdpwWpTxOC9:-KkWuJ5ifRdpwWpTxOCA:b-chwj26) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/automotive_demo.cc#L74)):*
> ```cpp
>               "dragway_base_speed + 2 * dragway_lane_speed_delta m/s.");
> DEFINE_double(dragway_vehicle_spacing, 10,
>               "The initial spacing (in meters) between consecutive vehicles"
> ```

BTW, nit, need a space at the end of this string, i.e. `... vehicles "`

---

*[drake/automotive/curve2.h, line 17 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW6OJDdD7e-R_kJZjd:-KkW6OJDdD7e-R_kJZje:b-1s2pql) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/curve2.h#L17)):*
> ```objc
> /// Curve2 represents a path through two-dimensional Cartesian space. Given a
> /// list of waypoints, it traces a path between them.
> ///
> ```

FYI, please consider adding the following documentation here:
```
/// Instantiated templates for the following kinds of T's are provided:
/// - double
/// - drake::AutoDiffXd
```

---

*[drake/automotive/curve2.h, line 41 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW8MKh2XnKZO0j_Iig:-KkW8MKh2XnKZO0j_Iih:bo2k2ej) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/curve2.h#L41)):*
> ```objc
> 
>   /// @return the waypoints associated with this curve.
>   std::vector<Point2> waypoints() const { return waypoints_; }
> ```

BTW, please consider the following:

```c++
 const std::vector<Point2>& waypoints() const { return waypoints_; }
```

as long as in our use cases it's always the case that the lifetime of the reference does not exceed the lifetime of the `Curve2<T>` which returned it.

---

*[drake/automotive/trajectory_car.h, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW9Z3cEmi7zFcQ0Pkv:-KkW9Z3cEmi7zFcQ0Pkw:b-1s2pql) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L52)):*
> ```objc
> /// output port 2: A FrameVelocity containing Xdot_WC, where C is the car frame.
> ///   (OutputPortDescriptor getter: velocity_output())
> ///
> ```

FYI, please consider adding the following documentation here:
```
/// Instantiated templates for the following kinds of T's are provided:
/// - double
/// - drake::AutoDiffXd
```

---

*[drake/automotive/trajectory_car.h, line 63 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWGGkCjIuHBqA4fYWk:-KkWGGkCjIuHBqA4fYWl:b-51pkx0) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L63)):*
> ```objc
>   /// Constructs a TrajectoryCar system that traces a given two-dimensional @p
>   /// curve.  Throws an error if the curve is empty (has a zero @p path_length).
>   explicit TrajectoryCar(const Curve2<T>& curve) : curve_(curve) {
> ```

FYI, consider the following:

```c++
#include <utility>  // needed because of std::move
...
explicit TrajectoryCar(Curve2<T> curve) : curve_(std::move(curve)) {
```

 - This version is as efficient as the previous one if `curve` is an lvalue. 
 - When `curve` is an rvalue, we have one move operation which should be more efficient than a copy operation. I've checked that `Curve2<T>` is move-constructible.

---

*[drake/automotive/trajectory_car.h, line 248 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWFeAU_YBvC9PtGqwV:-KkWFeAU_YBvC9PtGqwW:b9ujoy) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L248)):*
> ```objc
> 
>   TrajectoryCar<AutoDiffXd>* DoToAutoDiffXd() const override {
>     std::vector<typename Curve2<T>::Point2> waypoints = curve_.waypoints();
> ```

BTW, please consider the following if you plan to change the return type of `Curve2::waypoints()` (see Curve2.h file below):

```c++
const std::vector<typename Curve2<T>::Point2>& waypoints = curve_.waypoints();
```

---

*[drake/automotive/trajectory_car.h, line 250 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWFUl9f8Rq5m3ASAn-:-KkWFUl9f8Rq5m3ASAn0:b-crs21a) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L250)):*
> ```objc
>     std::vector<typename Curve2<T>::Point2> waypoints = curve_.waypoints();
>     std::vector<Curve2<AutoDiffXd>::Point2> autodiff_waypoints{};
>     for (auto point : waypoints) {
> ```

BTW, `auto point` -> `const auto& point`

---

*[drake/automotive/trajectory_car.h, line 251 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWFDjVsNA032-NxDMc:-KkWFDjVsNA032-NxDMd:b3yilsd) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L251)):*
> ```objc
>     std::vector<Curve2<AutoDiffXd>::Point2> autodiff_waypoints{};
>     for (auto point : waypoints) {
>       autodiff_waypoints.emplace_back(Curve2<AutoDiffXd>::Point2(point));
> ```

FYI, I think the following is more concise and (possibly) efficient :
```
      autodiff_waypoints.emplace_back(point);
```

---

*[drake/automotive/trajectory_car.h, line 254 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWpYbjGSUCHBLS2uxU:-KkWpYbkakijGsJ0uySp:b-vrku4b) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L254)):*
> ```objc
>     }
>     const Curve2<AutoDiffXd> curve(autodiff_waypoints);
>     return new TrajectoryCar<AutoDiffXd>(curve);
> ```

and merge the two lines into
```c++
    return new TrajectoryCar<AutoDiffXd>(
        Curve2<AutoDiffXd>(autodiff_waypoints));
```
to take advantage of the suggested `TrajectoryCar` constructor above.

---

*[drake/automotive/test/curve2_test.cc, line 37 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW8tsFo28lPMI-8UUo:-KkW8tsFo28lPMI-8UUp:b-8c4a5p) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/test/curve2_test.cc#L37)):*
> ```cpp
>   const Curve2d segment{segment_waypoints};
>   EXPECT_DOUBLE_EQ(segment.path_length(), M_SQRT2);
>   auto waypoints = segment.waypoints();
> ```

FYI, please consider adding:

```c++
  ASSERT_GE(waypoints.size(), 2);
```

---

*[drake/automotive/test/trajectory_car_test.cc, line 21 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWqozDfMCEau8EMk2w:-KkWqozDfMCEau8EMk2x:b-bfvco5) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/test/trajectory_car_test.cc#L21)):*
> ```cpp
> using systems::rendering::PoseVector;
> 
> static void SetAcceleration(const TrajectoryCar<double>& car_dut,
> ```

BTW, `static` is not necessary here since we're in the `namespace {}`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-:-KkWwg-5YIT5JAQEaNKy:bwjauqw)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 20 of 25 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 74 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWuJ5ifRdpwWpTxOC9:-KkkWrP68bAp8o7qa7W4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/automotive_demo.cc#L74)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit, need a space at the end of this string, i.e. `... vehicles "`
</blockquote></details>

Done.

---

*[drake/automotive/curve2.h, line 17 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW6OJDdD7e-R_kJZjd:-KkkXSRI_HOcBYafvzJh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/curve2.h#L17)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, please consider adding the following documentation here:
```
/// Instantiated templates for the following kinds of T's are provided:
/// - double
/// - drake::AutoDiffXd
```

</blockquote></details>

Done.

---

*[drake/automotive/curve2.h, line 41 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW8MKh2XnKZO0j_Iig:-KkkcST5V183cZQ0ns7H:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/curve2.h#L41)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, please consider the following:

```c++
 const std::vector<Point2>& waypoints() const { return waypoints_; }
```

as long as in our use cases it's always the case that the lifetime of the reference does not exceed the lifetime of the `Curve2<T>` which returned it.

</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW9Z3cEmi7zFcQ0Pkv:-KkkcyajIQFIEnhg4eD7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L52)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, please consider adding the following documentation here:
```
/// Instantiated templates for the following kinds of T's are provided:
/// - double
/// - drake::AutoDiffXd
```

</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 63 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWGGkCjIuHBqA4fYWk:-KkkeZ8EQcGBQ0nBj0Ff:b-oj5lqw) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L63)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, consider the following:

```c++
#include <utility>  // needed because of std::move
...
explicit TrajectoryCar(Curve2<T> curve) : curve_(std::move(curve)) {
```

 - This version is as efficient as the previous one if `curve` is an lvalue. 
 - When `curve` is an rvalue, we have one move operation which should be more efficient than a copy operation. I've checked that `Curve2<T>` is move-constructible.


</blockquote></details>

Thanks for the pointer! Done.

---

*[drake/automotive/trajectory_car.h, line 248 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWFeAU_YBvC9PtGqwV:-Kkkeo_AK7THzgABb_-s:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L248)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, please consider the following if you plan to change the return type of `Curve2::waypoints()` (see Curve2.h file below):

```c++
const std::vector<typename Curve2<T>::Point2>& waypoints = curve_.waypoints();
```

</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 250 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWFUl9f8Rq5m3ASAn-:-Kkkeuv6QVsSbgh1F5H6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L250)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `auto point` -> `const auto& point`
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 251 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWFDjVsNA032-NxDMc:-Kkkf0DUekDT7wu_aT5d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L251)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, I think the following is more concise and (possibly) efficient :
```
      autodiff_waypoints.emplace_back(point);
```
</blockquote></details>

Done.

---

*[drake/automotive/trajectory_car.h, line 254 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWpYbjGSUCHBLS2uxU:-KkkfGYuLBUUEA4wYkBi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/trajectory_car.h#L254)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

and merge the two lines into
```c++
    return new TrajectoryCar<AutoDiffXd>(
        Curve2<AutoDiffXd>(autodiff_waypoints));
```
to take advantage of the suggested `TrajectoryCar` constructor above.
</blockquote></details>

Done.

---

*[drake/automotive/test/curve2_test.cc, line 37 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkW8tsFo28lPMI-8UUo:-KkkftKVNmbI0r43j4Ku:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/test/curve2_test.cc#L37)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, please consider adding:

```c++
  ASSERT_GE(waypoints.size(), 2);
```
</blockquote></details>

Done.

---

*[drake/automotive/test/trajectory_car_test.cc, line 21 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6119#-KkWqozDfMCEau8EMk2w:-KkkgF9tQFD_QNsjQUTd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d632dfee47d08199f58f65175bcc638261548359/drake/automotive/test/trajectory_car_test.cc#L21)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `static` is not necessary here since we're in the `namespace {}`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6119)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@jadecastro In this PR we changed `TrajectoryCar(const Curve2<double>& curve)` to  `TrajectoryCar(const Curve2<T>& curve)`.  In general, passing `T` type or `T`-based types to constructor cannot work, because it does not allow the optimization framework to establish the symbolic variables, nor set up the autodiff derivatives, etc.  Is there any problem if I revert the code to use `Curve2<double>` exclusively, even for the `AutoDiff` form?

jadecastro commented:
Yes, I believe it would be fine to revert that.  (I should have never let that constructor take a `const Curve2<T>` as its argument.)  As long as we maintain a `Curve2<T>` as its member field, I think we should be fine.

jwnimmer-tri commented:
Thanks.  I'll assign you review of #6600.
